src/libraries/System.Linq/src/System/Linq/SelectManyWithParameterHelpers.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+
+namespace System.Linq
+{
+    internal static class SelectManyWithParameterHelpers
+    {
+        public static IEnumerable<TResult> SelectMany<TSource, TResult, TParam1>(IEnumerable<TSource> src,
+            Func<TSource, TParam1, IEnumerable<TResult>> func, TParam1 param1)
+        {
+            if (src is TSource[] srcArray)
+            {
+                return SelectManyArray(srcArray, func, param1);
+            }
+
+            return SelectManyNonArray(src, func, param1);
+        }
+
+        private static IEnumerable<TResult> SelectManyArray<TSource, TResult, TParam1>(TSource[] src,
+            Func<TSource, TParam1, IEnumerable<TResult>> func, TParam1 param1)
+        {
+            foreach (var elem in src)
+            {
+                foreach (var subElem in func(elem, param1))
+                {
+                    yield return subElem;
+                }
+            }
+        }
+        private static IEnumerable<TResult> SelectManyNonArray<TSource, TResult, TParam1>(IEnumerable<TSource> src,
+            Func<TSource, TParam1, IEnumerable<TResult>> func, TParam1 param1)
+        {
+            foreach (var elem in src)
+            {
+                foreach (var subElem in func(elem, param1))
+                {
+                    yield return subElem;
+                }
+            }
+        }
+        public static IEnumerable<TResult> SelectMany<TSource, TResult, TParam1, TParam2>(IEnumerable<TSource> src,
+            Func<TSource, TParam1, TParam2, IEnumerable<TResult>> func, TParam1 param1, TParam2 param2)
+        {
+            if (src is TSource[] srcArray)
+            {
+                return SelectManyArray(srcArray, func, param1, param2);
+            }
+
+            return SelectManyNonArray(src, func, param1, param2);
+        }
+
+        private static IEnumerable<TResult> SelectManyNonArray<TSource, TResult, TParam1, TParam2>(IEnumerable<TSource> src,
+            Func<TSource, TParam1, TParam2, IEnumerable<TResult>> func, TParam1 param1, TParam2 param2)
+        {
+            foreach (var elem in src)
+            {
+                foreach (var subElem in func(elem, param1, param2))
+                {
+                    yield return subElem;
+                }
+            }
+        }
+        private static IEnumerable<TResult> SelectManyArray<TSource, TResult, TParam1, TParam2>(TSource[] src,
+            Func<TSource, TParam1, TParam2, IEnumerable<TResult>> func, TParam1 param1, TParam2 param2)
+        {
+            foreach (var elem in src)
+            {
+                foreach (var subElem in func(elem, param1, param2))
+                {
+                    yield return subElem;
+                }
+            }
+        }
```

Comment:
Why do we need to differentiate between array and non array?

Reply:
The data source is Array as of now so I want to provide an optimized path for it. Meanwhile it would still work if the source becomes non-array in the future.

Reply:
Where is the optimized path? The method bodies are identical.

Reply:
iirc `foreach` on arrays is compiled into `for(int i=0;i<...;i++)` while `foreach` on `(IEnumerable<T>)T[]` utilizes `SZGenericArrayEnumerator<T>` eventually.

Reply:
IMO such a perf optimization isn't worth the cost of having two code paths.